SU-HSDO/suhumsci
# HSD8-1252: Cesta publications importer

# READY FOR REVIEW
## DO NOT MERGE

## Summary
Created a CSV publications importer for CESTA.

## Steps to Test
1. Import the migration configuration locally on a copy of the CESTA site.
    - `/admin/config/development/configuration/single/import`
1. Download the "CESTA Publications Template - 2022-04-05 - downloaded and adjusted" file from the ticket
    - https://stanfordits.atlassian.net/browse/HSD8-1252
1. Import the CSV file.
    - `/admin/structure/migrate/manage/default/migrations/custm_cesta_publications/csv-upload`
    - There will be some encoding errors – 27 of 92 will fail due to an iconv() notice:
      - `Notice: iconv(): Detected an illegal character in input string in Drupal\stanford_migrate\Form\StanfordMigrateCsvImportForm->fixLineBreaks() (line 290 of modules/custom/stanford_migrate/src/Form/StanfordMigrateCsvImportForm.php).`
      - This can be fixed with cleaning up the bad characters.
1. Open some of the newly created publication nodes and verify the data.
    - **NOTE:** My local was caching the /admin/content view. I had to rebuild my cache to see the imported Publications.
1. Review instructions for user below.
    
## Instructions for use
1. Go to the CSV upload page for the migration
    1. `/admin/structure/migrate/manage/default/migrations/custm_cesta_publications/csv-upload`
1. Download the example CSV file
1. Paste your data into the file
1. Make sure every row in the file has a unique ID number
    1. The ID number is used to align data for subsequent imports. If you need to change the data in the sheet and re-import, make sure the ID's align. Otherwise it will create a new publication or update the wrong publication.
1. Clean up the data in the file
    1. The Author's and Co-Author's data should be in the form of "First Last" and can have multiple comma delimited `,` values.
    1. The Publication Type data should only have 1 value.
    1. The Year data should only contain the year – full dates aren't accepted.
    1. The DH Method or Approach data can have multiple comma delimited `,` values.
    1. Remove or change any special characters.
    1. Globally, no values are being ignored; if the data should not be imported, leave it blank. E.g., values like "n/a" or "n/a for DH method" will be imported and need to be changed to blank values.

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)
